# Omkar Biradar (0126)

# Import Libraries

In [204]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split, RandomizedSearchCV, GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

import warnings
warnings.filterwarnings("ignore")

# Data Gathering

In [205]:
df = pd.read_csv("Melbourne_housing_FULL.csv")
df

Suburb           Address  Rooms Type Method        SellerG  \
0            Abbotsford     68 Studley St      2    h     SS         Jellis   
1          Airport West     154 Halsey Rd      3    t     PI         Nelson   
2           Albert Park    105 Kerferd Rd      2    h      S  hockingstuart   
3           Albert Park  85 Richardson St      2    h      S        Thomson   
4            Alphington      30 Austin St      3    h     SN        McGrath   
...                 ...               ...    ...  ...    ...            ...   
34852         Reservoir      18 Elinda Pl      3    u     SP             RW   
34853     Roxburgh Park    14 Stainsby Cr      4    h      S          Raine   
34854  Springvale South     8 Bellbird Ct      4    h     PI          Barry   
34855  Springvale South  30 Waddington Cr      3    h      S      Harcourts   
34856       Westmeadows      42 Pascoe St      4    h      S          Barry   

             Date  Distance  Postcode  Bedroom  ...  Landsize  BuildingArea  \
0        3/9/2016       2.5    3067.0      2.0  ...     126.0           inf   
1        3/9/2016      13.5    3042.0      3.0  ...     303.0           225   
2        3/9/2016       3.3    3206.0      2.0  ...     120.0            82   
3        3/9/2016       3.3    3206.0      2.0  ...     159.0           inf   
4        3/9/2016       6.4    3078.0      3.0  ...     174.0           122   
...           ...       ...       ...      ...  ...       ...           ...   
34852  30/09/2017      12.0    3073.0      3.0  ...       NaN         105.0   
34853  30/09/2017      20.6    3064.0      4.0  ...       NaN         225.0   
34854  30/09/2017      22.2    3172.0      4.0  ...     534.0         152.0   
34855  30/09/2017      22.2    3172.0      3.0  ...     544.0           NaN   
34856  30/09/2017      16.5    3049.0      4.0  ...     813.0         140.0   

       YearBuilt                     CouncilArea  Latitude Longtitude  \
0            NaN              Yarra City Council -37.80140  144.99580   
1         2016.0      Moonee Valley City Council -37.71800  144.87800   
2         1900.0       Port Phillip City Council -37.84590  144.95550   
3            NaN       Port Phillip City Council -37.84500  144.95380   
4         2003.0            Darebin City Council -37.78180  145.01980   
...          ...                             ...       ...        ...   
34852     1990.0            Darebin City Council -37.69769  145.02332   
34853     1995.0               Hume City Council -37.63665  144.92976   
34854     1970.0  Greater Dandenong City Council -37.97037  145.15449   
34855        NaN  Greater Dandenong City Council -37.97751  145.14813   
34856     1960.0               Hume City Council -37.67631  144.89409   

                       Regionname  Propertycount      ParkingArea      Price  
0           Northern Metropolitan         4019.0          Carport        NaN  
1            Western Metropolitan         3464.0  Detached Garage   840000.0  
2           Southern Metropolitan         3280.0  Attached Garage  1275000.0  
3           Southern Metropolitan         3280.0           Indoor  1455000.0  
4           Northern Metropolitan         2211.0          Parkade        NaN  
...                           ...            ...              ...        ...  
34852       Northern Metropolitan        21650.0          Parkade   475000.0  
34853       Northern Metropolitan         5833.0      Underground   591000.0  
34854  South-Eastern Metropolitan         4054.0          Carport        NaN  
34855  South-Eastern Metropolitan         4054.0  Detached Garage   780500.0  
34856       Northern Metropolitan         2474.0  Attached Garage   791000.0  

[34857 rows x 22 columns]

In [206]:
df.shape

(34857, 22)

In [207]:
df.head()

Suburb           Address  Rooms Type Method        SellerG      Date  \
0    Abbotsford     68 Studley St      2    h     SS         Jellis  3/9/2016   
1  Airport West     154 Halsey Rd      3    t     PI         Nelson  3/9/2016   
2   Albert Park    105 Kerferd Rd      2    h      S  hockingstuart  3/9/2016   
3   Albert Park  85 Richardson St      2    h      S        Thomson  3/9/2016   
4    Alphington      30 Austin St      3    h     SN        McGrath  3/9/2016   

   Distance  Postcode  Bedroom  ...  Landsize  BuildingArea  YearBuilt  \
0       2.5    3067.0      2.0  ...     126.0           inf        NaN   
1      13.5    3042.0      3.0  ...     303.0           225     2016.0   
2       3.3    3206.0      2.0  ...     120.0            82     1900.0   
3       3.3    3206.0      2.0  ...     159.0           inf        NaN   
4       6.4    3078.0      3.0  ...     174.0           122     2003.0   

                  CouncilArea  Latitude Longtitude             Regionname  \
0          Yarra City Council  -37.8014   144.9958  Northern Metropolitan   
1  Moonee Valley City Council  -37.7180   144.8780   Western Metropolitan   
2   Port Phillip City Council  -37.8459   144.9555  Southern Metropolitan   
3   Port Phillip City Council  -37.8450   144.9538  Southern Metropolitan   
4        Darebin City Council  -37.7818   145.0198  Northern Metropolitan   

   Propertycount      ParkingArea      Price  
0         4019.0          Carport        NaN  
1         3464.0  Detached Garage   840000.0  
2         3280.0  Attached Garage  1275000.0  
3         3280.0           Indoor  1455000.0  
4         2211.0          Parkade        NaN  

[5 rows x 22 columns]

In [208]:
df.tail()

Suburb           Address  Rooms Type Method    SellerG  \
34852         Reservoir      18 Elinda Pl      3    u     SP         RW   
34853     Roxburgh Park    14 Stainsby Cr      4    h      S      Raine   
34854  Springvale South     8 Bellbird Ct      4    h     PI      Barry   
34855  Springvale South  30 Waddington Cr      3    h      S  Harcourts   
34856       Westmeadows      42 Pascoe St      4    h      S      Barry   

             Date  Distance  Postcode  Bedroom  ...  Landsize  BuildingArea  \
34852  30/09/2017      12.0    3073.0      3.0  ...       NaN         105.0   
34853  30/09/2017      20.6    3064.0      4.0  ...       NaN         225.0   
34854  30/09/2017      22.2    3172.0      4.0  ...     534.0         152.0   
34855  30/09/2017      22.2    3172.0      3.0  ...     544.0           NaN   
34856  30/09/2017      16.5    3049.0      4.0  ...     813.0         140.0   

       YearBuilt                     CouncilArea  Latitude Longtitude  \
34852     1990.0            Darebin City Council -37.69769  145.02332   
34853     1995.0               Hume City Council -37.63665  144.92976   
34854     1970.0  Greater Dandenong City Council -37.97037  145.15449   
34855        NaN  Greater Dandenong City Council -37.97751  145.14813   
34856     1960.0               Hume City Council -37.67631  144.89409   

                       Regionname  Propertycount      ParkingArea     Price  
34852       Northern Metropolitan        21650.0          Parkade  475000.0  
34853       Northern Metropolitan         5833.0      Underground  591000.0  
34854  South-Eastern Metropolitan         4054.0          Carport       NaN  
34855  South-Eastern Metropolitan         4054.0  Detached Garage  780500.0  
34856       Northern Metropolitan         2474.0  Attached Garage  791000.0  

[5 rows x 22 columns]

In [209]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34857 entries, 0 to 34856
Data columns (total 22 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         34857 non-null  object 
 1   Address        34857 non-null  object 
 2   Rooms          34857 non-null  int64  
 3   Type           34857 non-null  object 
 4   Method         34857 non-null  object 
 5   SellerG        34857 non-null  object 
 6   Date           34857 non-null  object 
 7   Distance       34856 non-null  float64
 8   Postcode       34856 non-null  float64
 9   Bedroom        26640 non-null  float64
 10  Bathroom       26631 non-null  float64
 11  Car            26129 non-null  float64
 12  Landsize       23047 non-null  float64
 13  BuildingArea   13760 non-null  object 
 14  YearBuilt      15551 non-null  float64
 15  CouncilArea    34854 non-null  object 
 16  Latitude       26881 non-null  float64
 17  Longtitude     26881 non-null  float64
 18  Region

In [210]:
df.describe()

Rooms      Distance      Postcode       Bedroom      Bathroom  \
count  34857.000000  34856.000000  34856.000000  26640.000000  26631.000000   
mean       3.031012     11.184929   3116.062859      3.084647      1.624798   
std        0.969933      6.788892    109.023903      0.980690      0.724212   
min        1.000000      0.000000   3000.000000      0.000000      0.000000   
25%        2.000000      6.400000   3051.000000      2.000000      1.000000   
50%        3.000000     10.300000   3103.000000      3.000000      2.000000   
75%        4.000000     14.000000   3156.000000      4.000000      2.000000   
max       16.000000     48.100000   3978.000000     30.000000     12.000000   

                Car       Landsize     YearBuilt      Latitude    Longtitude  \
count  26129.000000   23047.000000  15551.000000  26881.000000  26881.000000   
mean       1.728845     593.598993   1965.289885    -37.810634    145.001851   
std        1.010771    3398.841946     37.328178      0.090279      0.120169   
min        0.000000       0.000000   1196.000000    -38.190430    144.423790   
25%        1.000000     224.000000   1940.000000    -37.862950    144.933500   
50%        2.000000     521.000000   1970.000000    -37.807600    145.007800   
75%        2.000000     670.000000   2000.000000    -37.754100    145.071900   
max       26.000000  433014.000000   2106.000000    -37.390200    145.526350   

       Propertycount         Price  
count   34854.000000  2.724700e+04  
mean     7572.888306  1.050173e+06  
std      4428.090313  6.414671e+05  
min        83.000000  8.500000e+04  
25%      4385.000000  6.350000e+05  
50%      6763.000000  8.700000e+05  
75%     10412.000000  1.295000e+06  
max     21650.000000  1.120000e+07

# Feature Engineering

In [211]:
df

Suburb           Address  Rooms Type Method        SellerG  \
0            Abbotsford     68 Studley St      2    h     SS         Jellis   
1          Airport West     154 Halsey Rd      3    t     PI         Nelson   
2           Albert Park    105 Kerferd Rd      2    h      S  hockingstuart   
3           Albert Park  85 Richardson St      2    h      S        Thomson   
4            Alphington      30 Austin St      3    h     SN        McGrath   
...                 ...               ...    ...  ...    ...            ...   
34852         Reservoir      18 Elinda Pl      3    u     SP             RW   
34853     Roxburgh Park    14 Stainsby Cr      4    h      S          Raine   
34854  Springvale South     8 Bellbird Ct      4    h     PI          Barry   
34855  Springvale South  30 Waddington Cr      3    h      S      Harcourts   
34856       Westmeadows      42 Pascoe St      4    h      S          Barry   

             Date  Distance  Postcode  Bedroom  ...  Landsize  BuildingArea  \
0        3/9/2016       2.5    3067.0      2.0  ...     126.0           inf   
1        3/9/2016      13.5    3042.0      3.0  ...     303.0           225   
2        3/9/2016       3.3    3206.0      2.0  ...     120.0            82   
3        3/9/2016       3.3    3206.0      2.0  ...     159.0           inf   
4        3/9/2016       6.4    3078.0      3.0  ...     174.0           122   
...           ...       ...       ...      ...  ...       ...           ...   
34852  30/09/2017      12.0    3073.0      3.0  ...       NaN         105.0   
34853  30/09/2017      20.6    3064.0      4.0  ...       NaN         225.0   
34854  30/09/2017      22.2    3172.0      4.0  ...     534.0         152.0   
34855  30/09/2017      22.2    3172.0      3.0  ...     544.0           NaN   
34856  30/09/2017      16.5    3049.0      4.0  ...     813.0         140.0   

       YearBuilt                     CouncilArea  Latitude Longtitude  \
0            NaN              Yarra City Council -37.80140  144.99580   
1         2016.0      Moonee Valley City Council -37.71800  144.87800   
2         1900.0       Port Phillip City Council -37.84590  144.95550   
3            NaN       Port Phillip City Council -37.84500  144.95380   
4         2003.0            Darebin City Council -37.78180  145.01980   
...          ...                             ...       ...        ...   
34852     1990.0            Darebin City Council -37.69769  145.02332   
34853     1995.0               Hume City Council -37.63665  144.92976   
34854     1970.0  Greater Dandenong City Council -37.97037  145.15449   
34855        NaN  Greater Dandenong City Council -37.97751  145.14813   
34856     1960.0               Hume City Council -37.67631  144.89409   

                       Regionname  Propertycount      ParkingArea      Price  
0           Northern Metropolitan         4019.0          Carport        NaN  
1            Western Metropolitan         3464.0  Detached Garage   840000.0  
2           Southern Metropolitan         3280.0  Attached Garage  1275000.0  
3           Southern Metropolitan         3280.0           Indoor  1455000.0  
4           Northern Metropolitan         2211.0          Parkade        NaN  
...                           ...            ...              ...        ...  
34852       Northern Metropolitan        21650.0          Parkade   475000.0  
34853       Northern Metropolitan         5833.0      Underground   591000.0  
34854  South-Eastern Metropolitan         4054.0          Carport        NaN  
34855  South-Eastern Metropolitan         4054.0  Detached Garage   780500.0  
34856       Northern Metropolitan         2474.0  Attached Garage   791000.0  

[34857 rows x 22 columns]

In [212]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34857 entries, 0 to 34856
Data columns (total 22 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         34857 non-null  object 
 1   Address        34857 non-null  object 
 2   Rooms          34857 non-null  int64  
 3   Type           34857 non-null  object 
 4   Method         34857 non-null  object 
 5   SellerG        34857 non-null  object 
 6   Date           34857 non-null  object 
 7   Distance       34856 non-null  float64
 8   Postcode       34856 non-null  float64
 9   Bedroom        26640 non-null  float64
 10  Bathroom       26631 non-null  float64
 11  Car            26129 non-null  float64
 12  Landsize       23047 non-null  float64
 13  BuildingArea   13760 non-null  object 
 14  YearBuilt      15551 non-null  float64
 15  CouncilArea    34854 non-null  object 
 16  Latitude       26881 non-null  float64
 17  Longtitude     26881 non-null  float64
 18  Region

In [213]:
df['Type'].value_counts().to_dict()

{'h': 23980, 'u': 7297, 't': 3580}

In [214]:
df['Type'].replace({'h':1,'u':2,'t':3},inplace=True)

# Handling Missing Values

In [215]:
df.isnull().sum()

Suburb               0
Address              0
Rooms                0
Type                 0
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom           8217
Bathroom          8226
Car               8728
Landsize         11810
BuildingArea     21097
YearBuilt        19306
CouncilArea          3
Latitude          7976
Longtitude        7976
Regionname           0
Propertycount        3
ParkingArea          0
Price             7610
dtype: int64

In [216]:
df["Distance"].unique()

array([ 2.5, 13.5,  3.3,  6.4, 13.8, 11.1,  6.3,  5.9, 11. , 12.2, 10.5,
        6.6,  9.7,  9.2, 13. , 13.9, 13.1, 10.8, 11.2, 10.7,  5.2, 11.8,
       11.7,  7.8,  9. ,  3.2, 11.4,  8.9,  8.1,  9.3, 13.6,  3.4,  1.6,
       10.3,  8.5,  7.7,  8. ,  9.4,  5.8,  3.5,  4.4, 12.1, 13.7, 14.5,
        4.6,  7.9, 15. , 12.8,  4.2,  5.6,  7.5,  6.2,  7.4,  8.7,  2.8,
        4.1,  6.9, 14.6,  8.4,  2.3,  5.5, 11.5, 14.7,  2.6,  9.9,  3.8,
        4.5,  8.8,  2.1,  1.2,  6.1, 12.6, 13.3,  6.5, 14.9,  1.9,  5.1,
        7. ,  9.1,  9.5,  1.8,  1.5,  9.8, 12.4, 10.6,  8.2, 10.4, 14. ,
       12.5,  5.3, 12.7, 10.9,  3. ,  5.7, 15.5,  4.3, 10.2, 16.5, 23.2,
       17.3, 35.2, 13.4, 25. , 16.1,  4. , 14.8, 10.1, 17.9, 17.5, 16.7,
        3.6, 43.4,  6.7, 20.6, 34.7, 23. , 24.7, 21.8, 22.2, 14.3, 25.9,
        1.3, 27. , 18. ,  7.2, 36.9, 19.6, 28.8, 24.8, 15.4, 38. , 45.9,
       12.9,  7.3, 20.4, 32.3, 45.2, 21.3, 16. , 18.4, 12.3,  5.4, 25.2,
       16.3, 33.8, 31.7, 29.8, 20. , 17.2, 16.2, 21

In [217]:
df['Distance'].mean()

11.18492942391554

In [218]:
df['Distance'].fillna(df['Distance'].mean(), inplace=True)

In [219]:
df['Distance'].isna().sum()

0

In [220]:
df['Postcode'].unique()

array([3067., 3042., 3206., 3078., 3018., 3025., 3143., 3032., 3147.,
       3034., 3183., 3103., 3104., 3204., 3165., 3128., 3019., 3186.,
       3187., 3056., 3055., 3105., 3125., 3124., 3126., 3054., 3163.,
       3162., 3161., 3148., 3068., 3058., 3066., 3108., 3084., 3185.,
       3184., 3040., 3041., 3065., 3031., 3011., 3146., 3046., 3043.,
       3188., 3122., 3081., 3166., 3079., 3021., 3033., 3101., 3102.,
       3144., 3012., 3145., 3000., 3127., 3039., 3189., 3015., 3051.,
       3070., 3167., 3052., 3044., 3207., 3181., 3072., 3073., 3121.,
       3205., 3141., 3006., 3182., 3020., 3107., 3071., 3142., 3087.,
       3003., 3016., 3085., 3013., 3057., 3061., 3053., 3002., 3060.,
       3123., 3047., 3083., 3008., 3028., 3022., 3049., 3153., 3193.,
       3806., 3130., 3155., 3088., 3023., 3151., 3192., 3169., 3168.,
       3978., 3064., 3977., 3136., 3175., 3089., 3172., 3109., 3111.,
       3754., 3196., 3095., 3782., 3076., 3177., 3156., 3131., 3199.,
       3200., 3437.,

In [221]:
df["Postcode"].value_counts()

Postcode
3073.0    844
3046.0    638
3020.0    617
3121.0    612
3165.0    583
         ... 
3770.0      1
3428.0      1
3139.0      1
3786.0      1
3793.0      1
Name: count, Length: 211, dtype: int64

In [222]:
df['Postcode'] = df['Postcode'].fillna(3073.0)

In [223]:
df['Postcode'].isna().sum()

0

In [224]:
df['Bedroom'].unique()

array([ 2.,  3.,  5., nan,  4.,  1.,  6.,  7.,  0., 20.,  8., 30.,  9.,
       12., 10., 16.])

In [225]:
df['Bedroom'].value_counts()

Bedroom
3.0     11881
4.0      6348
2.0      5777
5.0      1427
1.0       966
6.0       168
7.0        30
0.0        17
8.0        13
9.0         5
10.0        4
20.0        1
30.0        1
12.0        1
16.0        1
Name: count, dtype: int64

In [226]:
df['Bedroom'] = df['Bedroom'].fillna(3.0)

In [227]:
df['Bedroom'].isna().sum()

0

In [228]:
df['Bathroom'].unique()

array([ 1.,  2.,  3., nan,  4.,  5.,  0., 12.,  7.,  8.,  6.,  9.])

In [229]:
df['Bathroom'].value_counts()

Bathroom
1.0     12969
2.0     11064
3.0      2181
4.0       269
5.0        77
0.0        46
6.0        16
7.0         4
8.0         3
12.0        1
9.0         1
Name: count, dtype: int64

In [230]:
df['Bathroom'] = df['Bathroom'].fillna(1.0)

In [231]:
df['Bathroom'].isna().sum()

0

In [232]:
df['Car'].unique()

array([ 1.,  0.,  4.,  2.,  6., nan,  3.,  7.,  5.,  9.,  8., 10., 12.,
       11., 26., 18.])

In [233]:
df['Car'].value_counts()

Car
2.0     12214
1.0      9164
0.0      1631
3.0      1606
4.0      1161
5.0       151
6.0       140
7.0        25
8.0        23
10.0        6
9.0         3
11.0        2
12.0        1
26.0        1
18.0        1
Name: count, dtype: int64

In [234]:
df['Car'] = df['Car'].fillna(2.0)

In [235]:
df['Car'].isna().sum()

0

In [236]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34857 entries, 0 to 34856
Data columns (total 22 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         34857 non-null  object 
 1   Address        34857 non-null  object 
 2   Rooms          34857 non-null  int64  
 3   Type           34857 non-null  int64  
 4   Method         34857 non-null  object 
 5   SellerG        34857 non-null  object 
 6   Date           34857 non-null  object 
 7   Distance       34857 non-null  float64
 8   Postcode       34857 non-null  float64
 9   Bedroom        34857 non-null  float64
 10  Bathroom       34857 non-null  float64
 11  Car            34857 non-null  float64
 12  Landsize       23047 non-null  float64
 13  BuildingArea   13760 non-null  object 
 14  YearBuilt      15551 non-null  float64
 15  CouncilArea    34854 non-null  object 
 16  Latitude       26881 non-null  float64
 17  Longtitude     26881 non-null  float64
 18  Region

In [237]:
df['Landsize'].unique()

array([ 126.,  303.,  120., ..., 2193., 1357., 9833.])

In [238]:
df['Landsize'].mean()

593.598993361392

In [239]:
df['Landsize'].isna().sum()

11810

In [240]:
df['Landsize'].fillna(df['Landsize'].mean(), inplace = True)

In [241]:
df['Landsize'].isna().sum()

0

In [242]:
df['BuildingArea'].unique()

array(['inf', '225', '82', '122', '263', '242', '108', '251', '117',
       'missing', '76', '135', '399', '118', '103', '180', nan, '123',
       '218', '129', '167', '154', '275', '121', '146', '125', '255',
       '94', '75', '254', '156', '404', '240', '370', '268', '202', '203',
       '69', '140', '214', '253', '189', '215', '195', '96', '104', '100',
       '313', '144', '130', '64', '93', '106', '107', '110', '70', '132',
       '229', '51', '147', '113', '83', '56', '137', '85', '175', '3558',
       '170', '259', '265', '353', '138', '19', '116', '87', '74', '320',
       '300', '52', '210', '120', '86', '97', '152', '200', '14', '161',
       '128', '178', '185', '109', '53', '133', '115', '143', '150',
       '236', '276', '188', '179', '249', '141', '349', '192', '34', '73',
       '84', '81', '207', '50', '197', '264', '312', '235', '221', '260',
       '183', '160', '186', '78', '105', '145', '168', '62', '111', '220',
       '315', '181', '500', '61', '112', '420', '226

In [243]:
df['BuildingArea'].value_counts()

BuildingArea
120       177
110       152
100       148
130       145
115       139
         ... 
85.35       1
78.54       1
266.53      1
114.2       1
154.0       1
Name: count, Length: 994, dtype: int64

In [244]:
df['BuildingArea'].replace({'inf':120,'missing':110}, inplace = True)

In [245]:
df['BuildingArea']= df['BuildingArea'].fillna(120)

In [246]:
df['BuildingArea'].isna().sum()

0

In [247]:
df['BuildingArea'] = df['BuildingArea'].astype(float)
print(df.dtypes)

Suburb            object
Address           object
Rooms              int64
Type               int64
Method            object
SellerG           object
Date              object
Distance         float64
Postcode         float64
Bedroom          float64
Bathroom         float64
Car              float64
Landsize         float64
BuildingArea     float64
YearBuilt        float64
CouncilArea       object
Latitude         float64
Longtitude       float64
Regionname        object
Propertycount    float64
ParkingArea       object
Price            float64
dtype: object


In [248]:
df['YearBuilt'].unique()

array([  nan, 2016., 1900., 2003., 1930., 2013., 2015., 1950., 1965.,
       1970., 1964., 1910., 1960., 1980., 1920., 1985., 2000., 1974.,
       1997., 2008., 1940., 1915., 1998., 1905., 1957., 1890., 1935.,
       2004., 1989., 2009., 1955., 1993., 1830., 1975., 1925., 2010.,
       1880., 2007., 2001., 2005., 1904., 1971., 1995., 1990., 1983.,
       2012., 2014., 1857., 1978., 1889., 2017., 1999., 1977., 2006.,
       2011., 2002., 1938., 1919., 1969., 1863., 1893., 1945., 1912.,
       1968., 1972., 1943., 1951., 1922., 1996., 1992., 1958., 1962.,
       1967., 1934., 1887., 1976., 1973., 1994., 1982., 1907., 1918.,
       1986., 1953., 1928., 1948., 1931., 1941., 1949., 1988., 1991.,
       1924., 1923., 1897., 1984., 1987., 1856., 1860., 2106., 1902.,
       1966., 1961., 1929., 1952., 1885., 1881., 1870., 1959., 1926.,
       1939., 1903., 1911., 1906., 1868., 1963., 1927., 1956., 1979.,
       1884., 1917., 1898., 1888., 1954., 1909., 1894., 1854., 1946.,
       1937., 1871.,

In [249]:
df['YearBuilt'].value_counts()

YearBuilt
1970.0    1490
1960.0    1260
1950.0    1089
1980.0     726
1900.0     606
          ... 
1856.0       1
1871.0       1
2106.0       1
1884.0       1
1820.0       1
Name: count, Length: 160, dtype: int64

In [250]:
df['YearBuilt'] = df['YearBuilt'].fillna(1970.0)

In [251]:
df['YearBuilt'].isna().sum()

0

In [252]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34857 entries, 0 to 34856
Data columns (total 22 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         34857 non-null  object 
 1   Address        34857 non-null  object 
 2   Rooms          34857 non-null  int64  
 3   Type           34857 non-null  int64  
 4   Method         34857 non-null  object 
 5   SellerG        34857 non-null  object 
 6   Date           34857 non-null  object 
 7   Distance       34857 non-null  float64
 8   Postcode       34857 non-null  float64
 9   Bedroom        34857 non-null  float64
 10  Bathroom       34857 non-null  float64
 11  Car            34857 non-null  float64
 12  Landsize       34857 non-null  float64
 13  BuildingArea   34857 non-null  float64
 14  YearBuilt      34857 non-null  float64
 15  CouncilArea    34854 non-null  object 
 16  Latitude       26881 non-null  float64
 17  Longtitude     26881 non-null  float64
 18  Region

In [253]:
df['Price'].unique()

array([     nan,  840000., 1275000., ...,  354000.,  190000.,  780500.])

In [254]:
df['Price'].mean()

1050173.344955408

In [255]:
df['Price'].fillna(df['Price'].mean(), inplace = True)

In [256]:
df['Price'].isna().sum()

0

In [257]:
df['Propertycount'].unique()

array([ 4019.,  3464.,  3280.,  2211.,  5301.,  5132.,  4836.,  6567.,
        3052.,  2894.,  4502.,  2952.,  5682.,  7809.,  6795., 10969.,
        4605.,  3589., 10579.,  6938., 11918.,  7082.,  4480.,  5678.,
        8920.,  3265.,  3106.,  7822.,  2379.,  6923.,  5051.,  3582.,
        2954., 11204.,  3445.,  4553.,  9028.,  1651.,  4898.,  8989.,
        9264.,  1308.,  2970.,  5825.,  6244.,  3593.,  7570., 10412.,
        8870.,  1071.,  2606.,  5454.,  2356., 11308.,  2947.,  2674.,
        3145.,  5549.,  1202.,  5629.,  5263., 10331.,  2671.,   394.,
        3873.,  4675.,  8801.,  4918., 17496.,  2019.,  2079.,  6232.,
        2555.,  5498.,  2291.,  6821., 11364.,  2651.,  3692.,  2309.,
        7485.,  8648.,  7717., 14577., 21650., 14949.,  3540.,  2417.,
        5943., 14887.,  8400.,  1223., 13240.,  3284.,  3755.,  4217.,
        6763.,  5457.,  5420.,  7217.,  2698.,  2329.,  2230.,  6380.,
         802.,  4380.,  1369.,  6543.,  1543.,   790.,  5533.,   438.,
      

In [258]:
df['Propertycount'].mean()

7572.8883055029555

In [259]:
df['Propertycount'].fillna(df['Propertycount'].mean(), inplace = True)

In [260]:
df['Propertycount'].isna().sum()

0

In [261]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34857 entries, 0 to 34856
Data columns (total 22 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         34857 non-null  object 
 1   Address        34857 non-null  object 
 2   Rooms          34857 non-null  int64  
 3   Type           34857 non-null  int64  
 4   Method         34857 non-null  object 
 5   SellerG        34857 non-null  object 
 6   Date           34857 non-null  object 
 7   Distance       34857 non-null  float64
 8   Postcode       34857 non-null  float64
 9   Bedroom        34857 non-null  float64
 10  Bathroom       34857 non-null  float64
 11  Car            34857 non-null  float64
 12  Landsize       34857 non-null  float64
 13  BuildingArea   34857 non-null  float64
 14  YearBuilt      34857 non-null  float64
 15  CouncilArea    34854 non-null  object 
 16  Latitude       26881 non-null  float64
 17  Longtitude     26881 non-null  float64
 18  Region

In [262]:
df.isna().sum()

Suburb              0
Address             0
Rooms               0
Type                0
Method              0
SellerG             0
Date                0
Distance            0
Postcode            0
Bedroom             0
Bathroom            0
Car                 0
Landsize            0
BuildingArea        0
YearBuilt           0
CouncilArea         3
Latitude         7976
Longtitude       7976
Regionname          0
Propertycount       0
ParkingArea         0
Price               0
dtype: int64

In [263]:
df

Suburb           Address  Rooms  Type Method        SellerG  \
0            Abbotsford     68 Studley St      2     1     SS         Jellis   
1          Airport West     154 Halsey Rd      3     3     PI         Nelson   
2           Albert Park    105 Kerferd Rd      2     1      S  hockingstuart   
3           Albert Park  85 Richardson St      2     1      S        Thomson   
4            Alphington      30 Austin St      3     1     SN        McGrath   
...                 ...               ...    ...   ...    ...            ...   
34852         Reservoir      18 Elinda Pl      3     2     SP             RW   
34853     Roxburgh Park    14 Stainsby Cr      4     1      S          Raine   
34854  Springvale South     8 Bellbird Ct      4     1     PI          Barry   
34855  Springvale South  30 Waddington Cr      3     1      S      Harcourts   
34856       Westmeadows      42 Pascoe St      4     1      S          Barry   

             Date  Distance  Postcode  Bedroom  ...    Landsize  BuildingArea  \
0        3/9/2016       2.5    3067.0      2.0  ...  126.000000         120.0   
1        3/9/2016      13.5    3042.0      3.0  ...  303.000000         225.0   
2        3/9/2016       3.3    3206.0      2.0  ...  120.000000          82.0   
3        3/9/2016       3.3    3206.0      2.0  ...  159.000000         120.0   
4        3/9/2016       6.4    3078.0      3.0  ...  174.000000         122.0   
...           ...       ...       ...      ...  ...         ...           ...   
34852  30/09/2017      12.0    3073.0      3.0  ...  593.598993         105.0   
34853  30/09/2017      20.6    3064.0      4.0  ...  593.598993         225.0   
34854  30/09/2017      22.2    3172.0      4.0  ...  534.000000         152.0   
34855  30/09/2017      22.2    3172.0      3.0  ...  544.000000         120.0   
34856  30/09/2017      16.5    3049.0      4.0  ...  813.000000         140.0   

       YearBuilt                     CouncilArea  Latitude Longtitude  \
0         1970.0              Yarra City Council -37.80140  144.99580   
1         2016.0      Moonee Valley City Council -37.71800  144.87800   
2         1900.0       Port Phillip City Council -37.84590  144.95550   
3         1970.0       Port Phillip City Council -37.84500  144.95380   
4         2003.0            Darebin City Council -37.78180  145.01980   
...          ...                             ...       ...        ...   
34852     1990.0            Darebin City Council -37.69769  145.02332   
34853     1995.0               Hume City Council -37.63665  144.92976   
34854     1970.0  Greater Dandenong City Council -37.97037  145.15449   
34855     1970.0  Greater Dandenong City Council -37.97751  145.14813   
34856     1960.0               Hume City Council -37.67631  144.89409   

                       Regionname  Propertycount      ParkingArea  \
0           Northern Metropolitan         4019.0          Carport   
1            Western Metropolitan         3464.0  Detached Garage   
2           Southern Metropolitan         3280.0  Attached Garage   
3           Southern Metropolitan         3280.0           Indoor   
4           Northern Metropolitan         2211.0          Parkade   
...                           ...            ...              ...   
34852       Northern Metropolitan        21650.0          Parkade   
34853       Northern Metropolitan         5833.0      Underground   
34854  South-Eastern Metropolitan         4054.0          Carport   
34855  South-Eastern Metropolitan         4054.0  Detached Garage   
34856       Northern Metropolitan         2474.0  Attached Garage   

              Price  
0      1.050173e+06  
1      8.400000e+05  
2      1.275000e+06  
3      1.455000e+06  
4      1.050173e+06  
...             ...  
34852  4.750000e+05  
34853  5.910000e+05  
34854  1.050173e+06  
34855  7.805000e+05  
34856  7.910000e+05  

[34857 rows x 22 columns]

In [264]:
from sklearn.preprocessing import MinMaxScaler, StandardScaler

# Train Test Split

In [265]:
x = df.drop(['Suburb','Address','Method','SellerG','Date','Type','CouncilArea','Latitude','Longtitude','Regionname','ParkingArea'], axis=1)
y = df['Type']

In [266]:
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.2,random_state=45)

In [267]:
x_train.isna().sum()

Rooms            0
Distance         0
Postcode         0
Bedroom          0
Bathroom         0
Car              0
Landsize         0
BuildingArea     0
YearBuilt        0
Propertycount    0
Price            0
dtype: int64

In [268]:
y_train.value_counts()

Type
1    19186
2     5822
3     2877
Name: count, dtype: int64

In [269]:
std = StandardScaler()
std_array = std.fit_transform(x)
df_std = pd.DataFrame(std_array, columns=x.columns)
df_std

Rooms  Distance  Postcode   Bedroom  Bathroom       Car  \
0     -1.062988 -1.279322 -0.450020 -1.240788  -0.69549 -0.902356   
1     -0.031974  0.341018 -0.679333 -0.075393   0.76149 -0.902356   
2     -1.062988 -1.161479  0.824964 -1.240788  -0.69549 -2.034916   
3     -1.062988 -1.161479  0.824964 -1.240788  -0.69549 -2.034916   
4     -0.031974 -0.704838 -0.349122 -0.075393   0.76149 -0.902356   
...         ...       ...       ...       ...       ...       ...   
34852 -0.031974  0.120063 -0.394985 -0.075393  -0.69549 -0.902356   
34853  0.999040  1.386874 -0.477537  1.090003   0.76149  0.230203   
34854  0.999040  1.622560  0.513097  1.090003   0.76149  0.230203   
34855 -0.031974  1.622560  0.513097 -0.075393   0.76149 -0.902356   
34856  0.999040  0.782929 -0.615126  1.090003   0.76149  4.760440   

           Landsize  BuildingArea  YearBuilt  Propertycount     Price  
0     -1.691959e-01     -0.062790   0.083914      -0.802624  0.000000  
1     -1.051503e-01      0.352710   1.920851      -0.927968 -0.370593  
2     -1.713669e-01     -0.213161  -2.711424      -0.969523  0.396430  
3     -1.572552e-01     -0.062790   0.083914      -0.969523  0.713819  
4     -1.518276e-01     -0.054876   1.401717      -1.210950  0.000000  
...             ...           ...        ...            ...       ...  
34852  4.113642e-17     -0.122147   0.882583       3.179230 -1.014187  
34853  4.113642e-17      0.352710   1.082250      -0.392943 -0.809647  
34854 -2.156528e-02      0.063839   0.083914      -0.794720  0.000000  
34855 -1.794689e-02     -0.062790   0.083914      -0.794720 -0.475507  
34856  7.938800e-02      0.016353  -0.315420      -1.151553 -0.456993  

[34857 rows x 11 columns]

# Train Logistic Regression 

In [270]:
log = LogisticRegression()
log.fit(x_train,y_train)

LogisticRegression()

# Evaluation

In [271]:
#Testing data Analysis
y_pred_test = log.predict(x_test)

cnf_matrix = confusion_matrix(y_test,y_pred_test)
print("Confusion Matrix :\n",cnf_matrix)

acc_score_log = accuracy_score(y_test,y_pred_test)
print("Accuracy Score :\n",acc_score_log)

clf_report = classification_report(y_test,y_pred_test)
print("Classification Report :\n",clf_report)

Confusion Matrix :
 [[4513  281    0]
 [ 943  532    0]
 [ 633   70    0]]
Accuracy Score :
 0.7236087205966724
Classification Report :
               precision    recall  f1-score   support

           1       0.74      0.94      0.83      4794
           2       0.60      0.36      0.45      1475
           3       0.00      0.00      0.00       703

    accuracy                           0.72      6972
   macro avg       0.45      0.43      0.43      6972
weighted avg       0.64      0.72      0.67      6972



In [272]:
#Training data Analysis
y_pred_log = log.predict(x_train)

cnf_matrix = confusion_matrix(y_train,y_pred_log)
print("Confusion Matrix :\n",cnf_matrix)

acc_score_log = accuracy_score(y_train,y_pred_log)
print("Accuracy Score :\n",acc_score_log)

clf_report = classification_report(y_train,y_pred_log)
print("Classification Report :\n",clf_report)

Confusion Matrix :
 [[18030  1156     0]
 [ 3752  2070     0]
 [ 2625   252     0]]
Accuracy Score :
 0.7208176438945669
Classification Report :
               precision    recall  f1-score   support

           1       0.74      0.94      0.83     19186
           2       0.60      0.36      0.45      5822
           3       0.00      0.00      0.00      2877

    accuracy                           0.72     27885
   macro avg       0.44      0.43      0.42     27885
weighted avg       0.63      0.72      0.66     27885



# K- Nearest Neighbor

In [273]:
from sklearn.neighbors import KNeighborsClassifier

In [274]:
knn = KNeighborsClassifier()
knn.fit(x_train,y_train)

KNeighborsClassifier()

# Evaluation

In [275]:
#Testing data Analysis
y_pred = knn.predict(x_test)

cnf_matrix = confusion_matrix(y_test,y_pred)
print("Confusion Matrix :\n",cnf_matrix)

acc_score_knn = accuracy_score(y_test,y_pred)
print("Accuracy Score :\n",acc_score_knn)

clf_report = classification_report(y_test,y_pred)
print("Classification Report :\n",clf_report)

Confusion Matrix :
 [[4391  315   88]
 [ 719  723   33]
 [ 556  106   41]]
Accuracy Score :
 0.73938611589214
Classification Report :
               precision    recall  f1-score   support

           1       0.77      0.92      0.84      4794
           2       0.63      0.49      0.55      1475
           3       0.25      0.06      0.09       703

    accuracy                           0.74      6972
   macro avg       0.55      0.49      0.50      6972
weighted avg       0.69      0.74      0.70      6972



In [276]:
#Training data analysis
y_pred_knn = knn.predict(x_train)

cnf_matrix = confusion_matrix(y_train,y_pred_knn)
print("Confusion Matrix :\n",cnf_matrix)

acc_score_knn = accuracy_score(y_train,y_pred_knn)
print("Accuracy Score :\n",acc_score_knn)

clf_report = classification_report(y_train,y_pred_knn)
print("Classification Report :\n",clf_report)

Confusion Matrix :
 [[18199   801   186]
 [ 2211  3524    87]
 [ 1941   422   514]]
Accuracy Score :
 0.797453828223059
Classification Report :
               precision    recall  f1-score   support

           1       0.81      0.95      0.88     19186
           2       0.74      0.61      0.67      5822
           3       0.65      0.18      0.28      2877

    accuracy                           0.80     27885
   macro avg       0.74      0.58      0.61     27885
weighted avg       0.78      0.80      0.77     27885



In [277]:
std_scaler = StandardScaler()
std_scaler.fit(x_train)

StandardScaler()

In [278]:
std_scaler.fit_transform(x_train)

array([[-2.09936490e+00, -1.23273646e+00, -1.06492326e+00, ...,
         8.74089032e-02,  2.23974833e+00, -1.14862204e+00],
       [ 2.03984109e+00,  8.99635062e-01,  7.11329543e-01, ...,
         4.84689720e-01, -4.98616284e-01, -6.64865580e-04],
       [ 1.00503959e+00, -8.20968163e-01, -1.35381640e-01, ...,
        -1.58117053e+00,  6.22239470e-01,  2.69733113e+00],
       ...,
       [-2.09936490e+00, -7.03320079e-01,  6.19295719e-01, ...,
         8.74089032e-02, -1.04358019e+00, -1.22922061e+00],
       [-2.97619079e-02,  1.34922518e-01, -6.41567673e-01, ...,
        -7.07152730e-01, -1.12169018e+00, -4.21482833e-01],
       [-2.97619079e-02, -5.56259974e-01,  2.60363804e-01, ...,
         8.74089032e-02, -6.54610509e-01, -6.64865580e-04]])

In [279]:
std_scaler.transform(x_test)

array([[-2.97619079e-02, -7.76850131e-01, -7.70415027e-01, ...,
        -2.69355681e+00, -2.27488840e-01,  1.39202823e-01],
       [-1.06456341e+00, -4.82729922e-01, -3.37856053e-01, ...,
        -1.11231505e-01, -4.57303776e-01, -5.45884963e-01],
       [ 1.00503959e+00,  8.84929051e-01,  1.59126598e-01, ...,
         8.74089032e-02, -7.43782316e-01,  7.87495613e-01],
       ...,
       [-2.97619079e-02, -4.09199870e-01,  2.69567187e-01, ...,
        -2.69355681e+00,  2.76839812e-01, -6.64865580e-04],
       [-1.06456341e+00, -8.65086194e-01,  5.78893910e-02, ...,
         8.74089032e-02,  8.42798598e-01, -7.28107801e-01],
       [-2.97619079e-02, -2.91551786e-01,  2.78770569e-01, ...,
         8.74089032e-02,  6.40525334e-01,  1.81249908e+00]])

# Decision Tree Classifier

In [280]:
dt = DecisionTreeClassifier()
dt.fit(x_train,y_train)

DecisionTreeClassifier()

# Evaluation

In [281]:
#Testing data Analysis
y_pred = dt.predict(x_test)

cnf_matrix = confusion_matrix(y_test,y_pred)
print("Confusion Matrix :\n",cnf_matrix)

acc_score_dt = accuracy_score(y_test,y_pred)
print("Accuracy Score :\n",acc_score_dt)

clf_report = classification_report(y_test,y_pred)
print("Classification Report :\n",clf_report)

Confusion Matrix :
 [[4243  213  338]
 [ 228 1092  155]
 [ 286  140  277]]
Accuracy Score :
 0.8049340218014917
Classification Report :
               precision    recall  f1-score   support

           1       0.89      0.89      0.89      4794
           2       0.76      0.74      0.75      1475
           3       0.36      0.39      0.38       703

    accuracy                           0.80      6972
   macro avg       0.67      0.67      0.67      6972
weighted avg       0.81      0.80      0.81      6972



In [282]:
#Training data analysis
y_pred_dt = dt.predict(x_train)

cnf_matrix = confusion_matrix(y_train,y_pred_dt)
print("Confusion Matrix :\n",cnf_matrix)

acc_score_dt = accuracy_score(y_train,y_pred_dt)
print("Accuracy Score :\n",acc_score_dt)

clf_report = classification_report(y_train,y_pred_dt)
print("Classification Report :\n",clf_report)

Confusion Matrix :
 [[19132    24    30]
 [   81  5730    11]
 [  111    54  2712]]
Accuracy Score :
 0.9888470503855119
Classification Report :
               precision    recall  f1-score   support

           1       0.99      1.00      0.99     19186
           2       0.99      0.98      0.99      5822
           3       0.99      0.94      0.96      2877

    accuracy                           0.99     27885
   macro avg       0.99      0.97      0.98     27885
weighted avg       0.99      0.99      0.99     27885



# Hyper-parameter Tunning

In [283]:
dt = DecisionTreeClassifier()
hype = { 'criterion' : ['gini','entropy'],

        'max_depth' : np.arange(2,10),
        'min_samples_split' : np.arange(2,10),
        'min_samples_leaf' : np.arange(2,10),   
    }
rscv_model = RandomizedSearchCV(dt,hype,random_state=10,cv=4)
rscv_model.fit(x_train,y_train)

RandomizedSearchCV(cv=4, estimator=DecisionTreeClassifier(),
                   param_distributions={'criterion': ['gini', 'entropy'],
                                        'max_depth': array([2, 3, 4, 5, 6, 7, 8, 9]),
                                        'min_samples_leaf': array([2, 3, 4, 5, 6, 7, 8, 9]),
                                        'min_samples_split': array([2, 3, 4, 5, 6, 7, 8, 9])},
                   random_state=10)

In [284]:
rscv_model.best_estimator_

DecisionTreeClassifier(criterion='entropy', max_depth=9, min_samples_leaf=5,
                       min_samples_split=3)

In [285]:
y_pred_new = rscv_model.predict(x_test)

new_acc_score_dt = accuracy_score(y_test,y_pred_new)
print("New Accuracy Score for DT :\n",new_acc_score_dt)

New Accuracy Score for DT :
 0.8313253012048193


# Random Forest Classifier

In [286]:
rf = RandomForestClassifier()
rf.fit(x_train,y_train)

RandomForestClassifier()

# Evaluation

In [287]:
#Testing data analysis
y_pred_rf = rf.predict(x_test)

cnf_matrix = confusion_matrix(y_test,y_pred_rf)
print("Confusion Matrix :\n",cnf_matrix)

acc_score_rf = accuracy_score(y_test,y_pred_rf)
print("Accuracy Score :\n",acc_score_rf)

clf_report = classification_report(y_test,y_pred_rf)
print("Classification Report :\n",clf_report)

Confusion Matrix :
 [[4495  134  165]
 [ 189 1193   93]
 [ 282  157  264]]
Accuracy Score :
 0.8537005163511188
Classification Report :
               precision    recall  f1-score   support

           1       0.91      0.94      0.92      4794
           2       0.80      0.81      0.81      1475
           3       0.51      0.38      0.43       703

    accuracy                           0.85      6972
   macro avg       0.74      0.71      0.72      6972
weighted avg       0.84      0.85      0.85      6972



In [288]:
#Training data analysis
y_pred = rf.predict(x_train)

cnf_matrix = confusion_matrix(y_train,y_pred)
print("Confusion Matrix :\n",cnf_matrix)

acc_score_rf = accuracy_score(y_train,y_pred)
print("Accuracy score :\n",acc_score_rf)

clf_report = classification_report(y_train,y_pred)
print("Classification Report :\n",clf_report)

Confusion Matrix :
 [[19086    50    50]
 [   50  5741    31]
 [   85    45  2747]]
Accuracy score :
 0.9888470503855119
Classification Report :
               precision    recall  f1-score   support

           1       0.99      0.99      0.99     19186
           2       0.98      0.99      0.98      5822
           3       0.97      0.95      0.96      2877

    accuracy                           0.99     27885
   macro avg       0.98      0.98      0.98     27885
weighted avg       0.99      0.99      0.99     27885



# Hyper-parameter Tunning 

In [289]:
rf = RandomForestClassifier()
hype = { 'criterion' : ['gini','entropy'],
        'max_depth' : np.arange(2,10),
       'min_samples_split' : np.arange(2,10),
       'min_samples_leaf' : np.arange(2,10),
       'n_estimators' : np.arange(60,100)
       
       }
rscv = RandomizedSearchCV(rf,hype,random_state=10,cv=5)
rscv.fit(x_train,y_train)

RandomizedSearchCV(cv=5, estimator=RandomForestClassifier(),
                   param_distributions={'criterion': ['gini', 'entropy'],
                                        'max_depth': array([2, 3, 4, 5, 6, 7, 8, 9]),
                                        'min_samples_leaf': array([2, 3, 4, 5, 6, 7, 8, 9]),
                                        'min_samples_split': array([2, 3, 4, 5, 6, 7, 8, 9]),
                                        'n_estimators': array([60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76,
       77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93,
       94, 95, 96, 97, 98, 99])},
                   random_state=10)

In [290]:
rscv.best_estimator_

RandomForestClassifier(criterion='entropy', max_depth=9, min_samples_leaf=7,
                       min_samples_split=3, n_estimators=81)

In [291]:
y_pred_RF = rscv.predict(x_test)

new_acc_score_RF = accuracy_score(y_test,y_pred_RF)
print("New Accuracy Score for RF :\n",new_acc_score_RF)

New Accuracy Score for RF :
 0.8387837062535858


# By appling the classification model i.e. Logistic Regression, K-Nearest Neighbors, Decision Tree and Random Forest on the given dataset , we conclude that Radom Forest has the higest accuracy score of 83% among all the applied algorithm